### PR TITLE
Fixes a whole bunch of issues

### PR DESCRIPTION
**Pull request by psy0rz**
Its a xmas miracle! \o/

Fixed debugging, logging, exception handling and documentation.

Added support for multiple modbus function codes and read lengths.

Added parity/databits/stopbits settings.

### CODE REVIEW

**Comment by thompcd** [comment]
@psy0rz are you using this in conjunction with the instathings ecosystem, or are you running this locally? I'm trying to find something that will convert modbus to mqtt messages and map the registers to objects using config data. As far as I can tell, this is exactly what I'm looking for. Is my understanding correct?

**Comment by psy0rz** [comment]
Yes, i'm using it standalone.

And yes: If there exists a herdsman-converter for your device, it will regulary poll the device and convert it to nice usabe json messages via mqtt. (https://github.com/Instathings/modbus-herdsman-converters/tree/master/devices)

note that due to a bug you can configure it to poll just one modbus device. (issue #13 )

**Comment by thompcd** [comment]
> Yes, i'm using it standalone.
> 
> And yes: If there exists a herdsman-converter for your device, it will regulary poll the device and convert it to nice usabe json messages via mqtt. (https://github.com/Instathings/modbus-herdsman-converters/tree/master/devices)
> 
> note that due to a bug you can configure it to poll just one modbus device. (issue #13 )

Thanks for the confirmation. I've got some holiday time this week, I'm going to try adding some setting it up and adding some new herdsmen devices!

**Comment by psy0rz** [comment]
I'm always using influxdb + grafana for stats and to make cool dashboards.

**Comment by danielvandenberg95** [comment]
Have you looked into using a worker?
https://github.com/yaacov/node-modbus-serial/tree/a0a020fa6c0888ffa0b60a845e70b4dfc14d14ec/worker
Requires a newer version of modbus-serial, but gets rid of your switch case.

File: lib/events/onMQTTMessage.js [hunk 1]
```diff
@@ -8,50 +8,54 @@ const handler = Handler.getInstance();
...
-      const instance = handler.get(id);
-      if (instance) {
-        instance.remove();
+  try {
+    const {topic, message} = payload;
+    logger.debug(`Received MQTT message on '${topic}' with data '${message}'`);
+    switch (topic) {
+      case `${baseTopic}/configure/set`: {
+        const parsedMessage = JSON.parse(message);
+        // console.log(parsedMessage);
```
Comment (danielvandenberg95) [inline lib/events/onMQTTMessage.js:17]: Should this be debug or removed?

File: lib/methods/start.js [hunk 2]
```diff
@@ -18,19 +19,30 @@ module.exports = async function start() {
     const { devices, modbus } = settings.get();
     const options = {
       baudRate: modbus.baud_rate,
+      parity: modbus.parity ? modbus.parity : 'none',
```
Comment (danielvandenberg95) [inline lib/methods/start.js:22]: I feel like defaults are handled differently elsewhere. Also, couldn't this be shorter with ```modbus.parity || 'none'```? I don't remember how that handled 0.

File: lib/methods/start.js [hunk 3]
```diff
@@ -18,19 +19,30 @@ module.exports = async function start() {
...
+      dataBits: modbus.data_bits ? modbus.data_bits : 8,
+      stopBits: modbus.stop_bits ? modbus.stop_bits : 1,
     };
     await this.modbus.connectRTUBuffered(modbus.port, options);
+    this.modbus.setTimeout(1000);
```
Comment (danielvandenberg95) [inline lib/methods/start.js:27]: Shouldn't this be awaited? And potentially a setting? (I know, I also made a PR with a magic number here O:) )

File: lib/modbus.js [hunk 4]
```diff
@@ -7,12 +7,15 @@ const logger = require('./util/logger');
...
   constructor(mqtt, modbus, device) {
     super();
+
     this.mqtt = mqtt;
     this.modbus = modbus;
     this.id = device.id;
     this.model = device.model;
     this.modbusId = device.modbus_id;
     this.descriptor = modbusHerdsmanConverters.findByModbusModel(this.model);
+    if (this.descriptor === undefined) throw (Error(`Device ${device.id} has unkown model: ${this.model}`));
```
Comment (danielvandenberg95) [inline lib/modbus.js:17]: YES PLEASE! Would've saved me two hours yesterday....

File: lib/modbus.js [hunk 5]
```diff
@@ -46,11 +49,31 @@ class Modbus extends events.EventEmitter {
...
         const address = _.get(addressDescriptor, 'address');
+        // mode function code (default fc4)
+        const fc = addressDescriptor.fc ? addressDescriptor.fc : 4;
+        // modbus read length (default 1)
+        const readlen = addressDescriptor.len ? addressDescriptor.len : 1;
         let value;
         try {
-          value = await this.modbus.readInputRegisters(address, 1);
+          logger.debug(`Polling modbus device ${this.id}: modbusId=${this.modbusId}, fc=${fc}, address=${address}, readlen=${readlen}`);
+          switch (fc) {
```
Comment (danielvandenberg95) [inline lib/modbus.js:59]: See https://github.com/yaacov/node-modbus-serial/tree/a0a020fa6c0888ffa0b60a845e70b4dfc14d14ec/worker, could leave this switch case to that library.

Comment (danielvandenberg95) [reply lib/modbus.js:59]: Potentially add "type" to the address descriptor?

**Review by danielvandenberg95 — COMMENTED** [review]
Overall, looking great! 9/10 would merge. (Can't, unfortunately...)

File: lib/modbus.js [hunk 6]
```diff
@@ -46,11 +49,31 @@ class Modbus extends events.EventEmitter {
...
+              break;
+            case 4:
+              value = await this.modbus.readInputRegisters(address, readlen);
+              break;
+            default:
+              logger.error(('Unknown function code'));
+          }
         } catch (err) {
-          logger.error(err);
+          logger.error(`While reading modbus: ${err.toString()}`);
```
Comment (danielvandenberg95) [inline lib/modbus.js:76]: modbus2mqtt:error 2022-01-15 09:35:05: While reading modbus: [object Object]

File: lib/modbus.js (hunk 5 above)
Comment (danielvandenberg95) [reply lib/modbus.js:59]: Hm, I'm also a bit confused about the readlen parameter. It seems that it's always doubled? Readlen of 4 gives me a buffer of 8, readlen 2 a buffer of 4, and readlen 1 a buffer of 2. Maybe describe that somewhere, if that's indeed how it's supposed to work?

File: lib/methods/start.js (hunk 2 above)
Comment (danielvandenberg95) [reply lib/methods/start.js:22]: Nevermind this, would go wrong with handling 0.

**Pull request closed without merging**